### PR TITLE
Lkt: complete typing of "if" expressions

### PR DESCRIPTION
**Pull request by pmderodat**

### CODE REVIEW

**Review by thvnx — APPROVED** [review]
Looks good!

File: contrib/lkt/language/parser.py [hunk 1]
```diff
@@ -763,17 +779,27 @@ def expected_type():
...
 
             lambda if_expr=T.IfExpr: If(
                 # If we're resolving the type of the condition expression, then
                 # the result type needs to be a boolean.
                 Entity == if_expr.cond_expr,
                 Self.bool_type,
 
-                # Else, we're in a branch, use the expected type
-                if_expr.expected_type
+                # Else, we're in a branch: use type info from teh whole "if"
```
Comment (raph-amiard) [inline contrib/lkt/language/parser.py:800]: Typo "teh"

File: contrib/lkt/language/parser.py [hunk 2]
```diff
@@ -3152,6 +3178,94 @@ class IfExpr(Expr):
     alternatives = Field(type=T.ElsifBranch.list)
     else_expr = Field(type=T.Expr)
 
+    @langkit_property(return_type=T.Expr.entity.array)
+    def cond_exprs():
+        """
+        All expression that compute a condition.
```
Comment (raph-amiard) [inline contrib/lkt/language/parser.py:3184]: All expression*s*

File: contrib/lkt/language/parser.py [hunk 3]
```diff
@@ -3152,6 +3178,94 @@ class IfExpr(Expr):
...
+        """
+        return (
+            Entity.cond_expr.singleton
+            .concat(Entity.alternatives.map(lambda alt: alt.cond_expr))
+        )
+
+    @langkit_property(return_type=T.Expr.entity.array)
+    def result_exprs():
+        """
+        All expression that compute a result.
```
Comment (raph-amiard) [inline contrib/lkt/language/parser.py:3194]: Same

File: contrib/lkt/language/parser.py [hunk 4]
```diff
@@ -3152,6 +3178,94 @@ class IfExpr(Expr):
...
+        """
+        return (
+            Entity.then_expr.singleton
+            .concat(Entity.alternatives.map(lambda alt: alt.then_expr))
+            .concat(Entity.else_expr.singleton)
+        )
+
+    @langkit_property()
+    def check_correctness_pre():
+        # Check that all expression used in conditions return bools
```
Comment (raph-amiard) [inline contrib/lkt/language/parser.py:3204]: Same

File: contrib/lkt/language/parser.py [hunk 5]
```diff
@@ -3152,6 +3178,94 @@ class IfExpr(Expr):
...
+        invalid_conds = Entity.cond_exprs.map(
```
Comment (raph-amiard) [inline contrib/lkt/language/parser.py:3205]: For the bool check, you're not supposed to need such a check here: The type errors should be sufficient, here you're duplicating the type checker's job in check correctness pre. 

For example, for the already existing check on the bool cond of the if branch, for this code:

```
@invalid val a = if 2 then 3 else 4
```

I get 

```
test.lkt:1:21: error: Mismatched types: expected `Bool`, got a number literal
1 | @invalid val a = if 2 then 3 else 4
  |                     ^
```

For the "alternatives return compatible types", I'm not sure exactly what's going on but I'm almost certain this shouldn't be a "pre" check if it needs to be a check.
Is  there a specific reason you added those checks here ?

Comment (raph-amiard) [reply contrib/lkt/language/parser.py:3205]: On second thought, the "alternative return compatible types" should be handled in `expected_type`, in a similar way to what is done for list elements. I tested this locally and it seems to work:

```
diff --git a/contrib/lkt/language/parser.py b/contrib/lkt/language/parser.py
index 50c974c51..af4425e9f 100644
--- a/contrib/lkt/language/parser.py
+++ b/contrib/lkt/language/parser.py
@@ -762,7 +762,11 @@ class Expr(LktNode):
                 Self.bool_type,

                 # Else, we're in a branch, use the expected type
-                if_expr.expected_type
+                if_expr.expected_type._or(
+                    # If there is no expected type, use the cftype of the then
+                    # branch.
+                    if_expr.then_expr.expr_context_free_type
+                )
             ),

             lambda err_on_null=T.ErrorOnNull: err_on_null.expected_type,
```

**Review by raph-amiard — CHANGES_REQUESTED** [review]

File: contrib/lkt/language/parser.py (hunk 5 above)
Comment (pmderodat) [reply contrib/lkt/language/parser.py:3205]: Ah right, I did not realize that this was duplicated work. TBH, the separation of work between the regular expr type checking framework and `check_correctness_pre` isn’t very clear to me yet. I’ll adjust, thanks.

Comment (pmderodat) [reply contrib/lkt/language/parser.py:3205]: Your tentative patch is an improvement, but I think we should do even better: regardless of the branch from which we call `expected_type`, we should try to use the context free type of all other branches. This is necessary to correctly infer types in cases such as:

```
val v = if true then [] elif false then [i] else []
```

File: contrib/lkt/language/parser.py [hunk 6]
```diff
@@ -3241,6 +3251,42 @@ class IfExpr(Expr):
...
+        )
+
+    @langkit_property(return_type=T.TypeDecl.entity)
+    def if_expected_type():
+        """
+        Helper to handle ``if`` expressions in ``Expr.expected_type``.
+        """
+        # Use the if expression context free type if available, otherwise
+        # fallback on its expected type.
+        return Entity.expr_context_free_type.then(
```
Comment (raph-amiard) [inline contrib/lkt/language/parser.py:3285]: you can use `._or` here

**Review by raph-amiard — CHANGES_REQUESTED** [review]

File: contrib/lkt/language/parser.py (hunk 6 above)
Comment (pmderodat) [reply contrib/lkt/language/parser.py]: I didn’t know about it, thanks! Done, now.

**Review by raph-amiard — APPROVED** [review]

**Pull request merged**